src/ansys/motorcad/core/rpc_client_core.py

```diff
@@ -207,19 +211,62 @@ def __init__(
 
     def __del__(self):
         """Close Motor-CAD when MotorCAD object leaves memory."""
-        if self._connected is True:
-            if (
-                (self.reuse_parallel_instances is False)
-                and (self._open_new_instance is True)
-                and (self._compatibility_mode is False)
-            ):
-                # Close Motor-CAD if not asked to keep open
-                try:
-                    self.quit()
-                except Exception:
-                    # Don't raise exception at this point
-                    # Motor-CAD might already have been closed by user
-                    pass
+        if self._close_motorcad_on_exit():
+            try:
+                self._quit()
+            except Exception:
+                # Don't raise exception at this point
+                # Motor-CAD might already have been closed by user
+                pass
+
+    def _close_motorcad_on_exit(self):
+        """Check whether to close Motor-CAD when MotorCAD object __del__ is called."""
+        if (
+            (self.reuse_parallel_instances is False)
+            and (self._open_new_instance is True)
+            and (self._compatibility_mode is False)
+        ):
+            # Local Motor-CAD has been launched by Python
+            return True
+        elif _HAS_PIM and pypim.is_configured():
+            # Always try to close Ansys Lab instance
+            return True
+        else:
+            return False
+
+    # Close Motor-CAD if not asked to keep open
+    def _launch_motorcad_local(self, port):
+        """Launch local Motor-CAD instance."""
+        if self._open_new_instance is True:
+            if port != -1:
+                self._port = int(port)
+
+            self._open_motor_cad_local()
+
+        else:  # (self._open_new_instance == False)
+            if port != -1:
+                self._port = int(port)
+
+            else:  # port is not defined
+                self._find_free_motor_cad()
+
+    def _launch_motorcad_remote(self):
+        """Launch Motor-CAD in Ansys Lab."""
+        pim = pypim.connect()
+
+        self.pim_instance = pim.create_instance(product_name="motorcad")
+        self.pim_instance.wait_for_ready()
+        # get ip and port for motorcad
+        address = self.pim_instance.services["http"].uri
+
+        ip = address.split(":")[0] + ":" + address.split(":")[1]
```

Comment:
This address split is potentially buggy and I believe is unnecessary, since in the end the request passed to `requests` is `address + "/jsonrpc"`. Moreover the gobal `SERVER_IP` will prevent to connect to multiple remote sessions and create unexpected bugs: in local we multiply sessions with many ports, but in a container environment we multiply sessions with many remote hosts.

I strongly suggest to make `url` a member of `MotorCADConnection` instead of just `ip`, which solves both issues: in `_launch_motorcad_local()`, `url` would take the value `"http://localhost:" + port + "/jsonrpc"`, in `_launch_motorcad_remote()`, `url` would take the value `pim_instance.services["http"].uri + "/jsonrpc"`. I think that even outside of PIM, like a custom scenario where a user manages the launching themselves through a VM, docker-compose or anything, they can simply provide the URL themselves without bothering about the whole ip/port separation.

Reply:
Thanks @plule-ansys . I've come across this issue before. I'm going to change this in another PR.

Reply:
@jgsdavies Ok! that was the only issue I saw in this PR